backend/api/Areas/Keycloak/Controllers/UserController.cs

```diff
@@ -1,96 +1,261 @@
+        /// <param name="keycloakAdmin"></param>
+        public UserController(ILogger<UserController> logger, IPimsService pimsService, IPimsAdminService pimsAdminService, IMapper mapper, IKeycloakAdmin keycloakAdmin)
         {
             _logger = logger;
+            _pimsService = pimsService;
             _pimsAdminService = pimsAdminService;
             _mapper = mapper;
-            _optionsKeycloak = optionsKeycloak.CurrentValue;
-            _requestClient = requestClient;
+            _keycloakAdmin = keycloakAdmin;
         }
         #endregion
 
         #region Endpoints
         /// <summary>
-        /// Fetch a list of users from keycloak.
+        /// Sync the user for the specified 'id' from keycloak with PIMS.
+        /// If the user does not exist in keycloak it will return a 400-BadRequest.
+        /// If the user does not exist in PIMS it will add it.
+        /// Also links the user to the appropriate groups it is a member of within keycloak.!--
+        /// If the group does not exist in PIMS it will add it.
         /// </summary>
+        /// <param name="id"></param>
+        /// <exception type="KeyNotFoundException">The user does not exist in keycloak.</exception>
+        /// <returns></returns>
+        [HttpPost("sync/{id}")]
+        public async Task<IActionResult> SyncUserAsync(Guid id)
+        {
+            var kuser = await _keycloakAdmin.GetUserAsync(id) ?? throw new KeyNotFoundException();
+            var kgroups = await _keycloakAdmin.GetUserGroupsAsync(id);
+
+            var euser = _pimsAdminService.User.Find(id);
+            if (euser == null)
+            {
+                // The user does not exist in PIMS, it needs to be added.
+                euser = _mapper.Map<Entity.User>(kuser);
+                euser.DisplayName = $"{kuser.LastName}, {kuser.FirstName}";
+                foreach (var group in kgroups)
+                {
+                    var erole = _pimsAdminService.Role.Find(group.Id);
+
+                    // If the role doesn't exist, create it.
+                    if (erole == null)
+                    {
+                        erole = _mapper.Map<Entity.Role>(group);
+                        _pimsAdminService.Role.AddOne(erole);
+                    }
+
+                    euser.Roles.Add(new Entity.UserRole(euser, erole));
+                }
+                _pimsAdminService.User.AddOne(euser);
+            }
+            else
+            {
+                // The user exists in PIMS, it only needs to be updated.
+                var roles = euser?.Roles.ToArray();
+                _mapper.Map(kuser, euser);
+                foreach (var group in kgroups)
+                {
+                    var erole = _pimsAdminService.Role.Find(group.Id);
+
+                    // If the role doesn't exist, create it.
+                    if (erole == null)
```

Comment:
this provides the rationale for my previous comment, the controllers may end up needing to do this a lot if they are responsible for checking the role status and selectively adding/updating.